File: chirp-runner/internal/server/sources.go
package server

import (
	"encoding/json"
	"net/http"
)

const (
	qwenModelsTag    = "chirp-models-v0.1.3"
	qwenModelFile    = "qwen3-tts-model.gguf"
	qwenCodecFile    = "qwen3-tts-codec.gguf"
	qwenModelBaseURL = "https://huggingface.co/thewh1teagle/qwen3-tts-gguf/resolve/main"
	kokoroModelsTag  = "kokoro-v1.0"
	kokoroModelDir   = "chirp-kokoro-models-kokoro-v1.0"
	kokoroBundleURL  = "https://huggingface.co/thewh1teagle/chirp-kokoro-models/resolve/main/chirp-kokoro-models-kokoro-v1.0.tar.gz"
	voicesCatalogURL = "https://raw.githubusercontent.com/thewh1teagle/chirp/main/chirp-desktop/src/assets/voices.json"
)

type modelSourceFile struct {
	Name string `json:"name"`
	URL  string `json:"url"`
}

type modelSource struct {
	ID          string            `json:"id"`
	Name        string            `json:"name"`
	Version     string            `json:"version"`
	Recommended bool              `json:"recommended,omitempty"`
	Size        string            `json:"size"`
	Description string            `json:"description"`
	Files       []modelSourceFile `json:"files,omitempty"`
	ArchiveURL  string            `json:"archive_url,omitempty"`
	Directory   string            `json:"directory,omitempty"`
}

type modelSourcesResponse struct {
	Runtimes     []modelSource `json:"runtimes"`
	VoicesURL    string        `json:"voices_url"`
	DefaultPaths []string      `json:"default_paths"`
}

func modelSources() modelSourcesResponse {
	return modelSourcesResponse{
		Runtimes: []modelSource{
			{
				ID:          "qwen",
				Name:        "Qwen",
				Version:     qwenModelsTag,
				Size:        "~900 MB",
				Description: "Voice cloning, multilingual synthesis, best quality on supported GPU hardware.",
				Files: []modelSourceFile{
					{Name: qwenModelFile, URL: qwenModelBaseURL + "/" + qwenModelFile},
					{Name: qwenCodecFile, URL: qwenModelBaseURL + "/" + qwenCodecFile},
				},
				Directory: "chirp-models-q5_0",
			},
			{
				ID:          "kokoro",
				Name:        "Kokoro",
				Version:     kokoroModelsTag,
				Recommended: true,
				Size:        "~336 MB",
				Description: "Fast local multi-voice speech with a lighter model bundle.",
				ArchiveURL:  kokoroBundleURL,
				Directory:   kokoroModelDir,
			},
		},
		VoicesURL: voicesCatalogURL,
		DefaultPaths: []string{
			"macOS: ~/Library/Application Support/com.thewh1teagle.chirp/models",
			"Windows: %LOCALAPPDATA%\\com.thewh1teagle.chirp\\models",
			"Linux: ~/.local/share/com.thewh1teagle.chirp/models",
		},
	}
}

func (s *Server) handleModelSources(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	_ = json.NewEncoder(w).Encode(modelSources())
}

File: chirp-server/src/server/sources.rs
use serde::Serialize;
use utoipa::ToSchema;

const QWEN_MODELS_TAG: &str = "chirp-models-v0.1.3";
const QWEN_MODEL_FILE: &str = "qwen3-tts-model.gguf";
const QWEN_CODEC_FILE: &str = "qwen3-tts-codec.gguf";
const QWEN_MODEL_BASE_URL: &str = "https://huggingface.co/thewh1teagle/qwen3-tts-gguf/resolve/main";
const KOKORO_MODELS_TAG: &str = "kokoro-v1.0";
const KOKORO_MODEL_DIR: &str = "chirp-kokoro-models-kokoro-v1.0";
const KOKORO_BUNDLE_URL: &str = "https://huggingface.co/thewh1teagle/chirp-kokoro-models/resolve/main/chirp-kokoro-models-kokoro-v1.0.tar.gz";
const VOICES_CATALOG_URL: &str = "https://raw.githubusercontent.com/thewh1teagle/chirp/main/chirp-desktop/src/assets/voices.json";

#[derive(Debug, Serialize, ToSchema)]
pub struct ModelSourceFile {
    name: &'static str,
    url: String,
}

#[derive(Debug, Serialize, ToSchema)]
pub struct ModelSource {
    id: &'static str,
    name: &'static str,
    version: &'static str,
    #[serde(skip_serializing_if = "is_false")]
    recommended: bool,
    size: &'static str,
    description: &'static str,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    files: Vec<ModelSourceFile>,
    #[serde(skip_serializing_if = "Option::is_none")]
    archive_url: Option<&'static str>,
    directory: &'static str,
}

#[derive(Debug, Serialize, ToSchema)]
pub struct ModelSourcesResponse {
    runtimes: Vec<ModelSource>,
    voices_url: &'static str,
    default_paths: Vec<&'static str>,
}

pub fn model_sources() -> ModelSourcesResponse {
    ModelSourcesResponse {
        runtimes: vec![
            ModelSource {
                id: "qwen",
                name: "Qwen",
                version: QWEN_MODELS_TAG,
                recommended: false,
                size: "~900 MB",
                description: "Voice cloning, multilingual synthesis, best quality on supported GPU hardware.",
                files: vec![
                    ModelSourceFile {
                        name: QWEN_MODEL_FILE,
                        url: format!("{QWEN_MODEL_BASE_URL}/{QWEN_MODEL_FILE}"),
                    },
                    ModelSourceFile {
                        name: QWEN_CODEC_FILE,
                        url: format!("{QWEN_MODEL_BASE_URL}/{QWEN_CODEC_FILE}"),
                    },
                ],
                archive_url: None,
                directory: "chirp-models-q5_0",
            },
            ModelSource {
                id: "kokoro",
                name: "Kokoro",
                version: KOKORO_MODELS_TAG,
                recommended: true,
                size: "~336 MB",
                description: "Fast local multi-voice speech with a lighter model bundle.",
                files: Vec::new(),
                archive_url: Some(KOKORO_BUNDLE_URL),
                directory: KOKORO_MODEL_DIR,
            },
        ],
        voices_url: VOICES_CATALOG_URL,
        default_paths: vec![
            "macOS: ~/Library/Application Support/com.thewh1teagle.chirp/models",
            "Windows: %LOCALAPPDATA%\\com.thewh1teagle.chirp\\models",
            "Linux: ~/.local/share/com.thewh1teagle.chirp/models",
        ],
    }
}

fn is_false(value: &bool) -> bool {
    !*value
}
